web/app/fr/gospeak/web/pages/orga/partners/contacts/ContactCtrl.scala

```diff
@@ -0,0 +1,132 @@
+    "partner" -> partnerId,
+    "first_name" -> firstName,
+    "last_name" -> lastName,
+    "email" -> emailAddress,
+    "description" -> optional(description)
+  )(Contact.Data.apply)(Contact.Data.unapply))
+
+
+  def list(group: Group.Slug, partner: Partner.Slug, params: Page.Params): Action[AnyContent] = SecuredAction.async { implicit req =>
+    (for {
+      groupElt <- OptionT(groupRepo.find(user, group))
+      partnerElt <- OptionT(partnerRepo.find(groupElt.id, partner))
+      contacts <- OptionT.liftF(contactRepo.list(partner, params))
+      b = listBreadcrumb(groupElt, partnerElt)
+    } yield Ok(html.list(groupElt, partnerElt, contacts)(b))).value.map(_.getOrElse(groupNotFound(group))).unsafeToFuture()
+  }
+
+  def create(group: Group.Slug, partner: Partner.Slug): Action[AnyContent] = SecuredAction.async { implicit req =>
+    createForm(group, partner, createForm).unsafeToFuture()
+  }
+
+  def doCreate(group: Group.Slug, partner: Partner.Slug): Action[AnyContent] = SecuredAction.async { implicit req =>
+    val now = Instant.now()
+    createForm.bindFromRequest.fold(
+      formWithErrors => createForm(group, partner, formWithErrors),
+      data => (for {
+        exists <- OptionT.liftF(contactRepo.exists(partner, data.email))
+        _ <- OptionT.liftF((!exists).toIO(DuplicateEmailException(data.email)))
+        contact <- OptionT.liftF(contactRepo.create(data, by, now))
+      } yield Redirect(ContactRoutes.detail(group, partner, contact.id))).value.map(_.getOrElse(partnerNotFound(group, partner)))
+        .recoverWith {
+          case _: DuplicateEmailException => createForm(group, partner, createForm.bindFromRequest().withError("email", "Email already exists"))
+          case NonFatal(e) => createForm(group, partner, createForm.bindFromRequest().withGlobalError(e.getMessage))
+        }
+    ).unsafeToFuture()
+  }
+
+
+  private def createForm(group: Group.Slug, partner: Partner.Slug, form: Form[Contact.Data])(implicit req: SecuredRequest[CookieEnv, AnyContent]): IO[Result] = {
+    (for {
+      groupElt <- OptionT(groupRepo.find(user, group))
+      partnerElt <- OptionT(partnerRepo.find(groupElt.id, partner))
+      b = listBreadcrumb(groupElt, partnerElt).add("New" -> routes.ContactCtrl.create(group, partner))
+      call = routes.ContactCtrl.doCreate(group, partner)
+    } yield Ok(html.create(groupElt, partnerElt, form, call)(b))).value.map(_.getOrElse(partnerNotFound(group, partner)))
+  }
+
+
+  def detail(group: Group.Slug, partner: Partner.Slug, contact: Contact.Id): Action[AnyContent] = SecuredAction.async { implicit req =>
+    (for {
+      groupElt <- OptionT(groupRepo.find(user, group))
+      (partnerElt, contactElt) <- OptionT(contactRepo.find(partner, contact))
+      users <- OptionT.liftF(userRepo.list(partnerElt.users))
+      b = breadcrumb(groupElt, partnerElt, contactElt)
+      edit = routes.ContactCtrl.edit(group, partner, contact)
```

Comment:
Urls can be computed inside the template, no need to do it here

Reply:
You need to show me how ^^ it will fix the previous point.